storage/reads/reader.go

```diff
@@ -694,3 +715,70 @@ func groupKeyForGroup(kv [][]byte, readSpec *influxdb.ReadSpec, bnds execute.Bou
 	}
 	return execute.NewGroupKey(cols, vs)
 }
+
+type tagKeysIterator struct {
+	ctx       context.Context
+	bounds    execute.Bounds
+	s         Store
+	readSpec  influxdb.ReadTagKeysSpec
+	predicate *datatypes.Predicate
+	alloc     *memory.Allocator
+}
+
+func (ti *tagKeysIterator) Do(f func(flux.Table) error) error {
+	src := ti.s.GetSource(
+		uint64(ti.readSpec.OrganizationID),
+		uint64(ti.readSpec.BucketID),
+	)
+
+	var req datatypes.TagKeysRequest
+	if any, err := types.MarshalAny(src); err != nil {
+		return err
+	} else {
+		req.TagsSource = any
+	}
+	req.Predicate = ti.predicate
+	req.Range.Start = int64(ti.bounds.Start)
+	req.Range.End = int64(ti.bounds.Stop)
+
+	rs, err := ti.s.TagKeys(ti.ctx, &req)
+	if err != nil {
+		return err
+	}
+	return ti.handleRead(f, rs)
+}
+
+func (ti *tagKeysIterator) handleRead(f func(flux.Table) error, rs storage.StringIterator) error {
+	builder := execute.NewColListTableBuilder(nil, ti.alloc)
+	valueIdx, err := builder.AddCol(flux.ColMeta{
+		Label: execute.DefaultValueColLabel,
+		Type:  flux.TString,
+	})
+	if err != nil {
+		return err
+	}
+	defer builder.ClearData()
+
+	for rs.Next() {
+		if err := builder.AppendString(valueIdx, rs.Value()); err != nil {
+			return err
+		}
+	}
+
+	// Construct the table and add to the reference count
+	// so we can free the table later.
+	tbl, err := builder.Table()
+	if err != nil {
+		return err
+	}
+	tbl.RefCount(1)
```

Comment:
Where exactly are table reference counts used?

Reply:
It's used by the table implementation for releasing the memory. We're not very good about it and we should probably update this API to be `Retain` and `Release` and for the default reference count to be 1, but it seemed like the default right now was 0.